### PR TITLE
Experiment with adding the Lint code to the prototype

### PR DESCRIPTION
**Pull request by punchagan**
Partially addresses https://github.com/tarides/infrastructure/issues/329

@shonfeder I haven't been able to work much on the API clean-up after our pairing session, looking through the `opam-repo-ci` lint code. But, sharing this since you mentioned you may be able to spend some time on this today.

### CODE REVIEW

File: bin/main.ml [hunk 1]
```diff
@@ -82,6 +90,14 @@ let pkg_term =
   let info = Arg.info [] ~doc:"Package name + version" in
   Arg.required (Arg.pos 0 (Arg.some Arg.string) None info)
 
+let lint_cmd =
+  let doc = "List the opam repository directory" in
```
Comment (shonfeder) [inline bin/main.ml]: Should this be

```suggestion
  let doc = "Lint the opam repository directory" in
```

**Review by shonfeder — APPROVED** [review]

File: bin/main.ml [hunk 2]
```diff
@@ -1,6 +1,14 @@
 open Cmdliner
 open Prototype
 
+let lint pkg local_repo_dir =
+  (match local_repo_dir with
+  | Some d ->
+      print_endline @@ Printf.sprintf "Linting %s in %s ..." pkg d;
+      Lint.run_lint pkg d
+  | None -> print_endline "No opam repository directory specified.");
```
Comment (shonfeder) [inline bin/main.ml]: Just to help us track followups:

```suggestion
  | None -> 
    (* TODO Add error handling *)
    print_endline "No opam repository directory specified.");
```

File: bin/main.ml [hunk 3]
```diff
@@ -1,6 +1,14 @@
...
+  ()
```
Comment (shonfeder) [inline bin/main.ml]: Looks like this can just be:

```suggestion
  match local_repo_dir with
  | Some d ->
      print_endline @@ Printf.sprintf "Linting %s in %s ..." pkg d;
      Lint.run_lint pkg d
  | None -> print_endline "No opam repository directory specified."
```

File: lib/dir_helpers.ml [hunk 4]
```diff
@@ -6,3 +6,16 @@ let create_temp_dir prefix =
...
+let remove_dir dir =
+  let rec remove_dir_rec dir =
+    let entries = Sys.readdir dir in
+    Array.iter
+      (fun entry ->
+        let entry = Filename.concat dir entry in
+        if Sys.is_directory entry then remove_dir_rec entry
+        else Unix.unlink entry)
+      entries;
+    Unix.rmdir dir
```
Comment (shonfeder) [inline lib/dir_helpers.ml]: Would it be possible to just use `Sys` rather than both `Sys` and `Unix` modules here? Seems like that would be preferable if so. (Same with other uses in this file, but that is out of scope for this PR.)

File: lib/lint.ml [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+
+let path_from_pkg ~repo_dir pkg =
+  repo_dir // "packages"
+  // OpamPackage.Name.to_string (OpamPackage.name pkg)
+  // OpamPackage.to_string pkg
+
+let run_lint pkg repo_dir =
+  let pkg = OpamPackage.of_string pkg in
+  let opam_path = path_from_pkg ~repo_dir pkg // "opam" in
+  (* NOTE: We read from channel instead of file to prevent name and version
```
Comment (shonfeder) [inline lib/lint.ml]: It took me a few moments to figure out how reading from a file could impact name and version fields, and then realized this is comment about the particular opam functions we are using, and not about reading from streams per se. Something like this would have made it clear to me:

```suggestion
  (* NOTE: We use OpamFile.OPAM.read_from_channel instead of OpamFile.OPAM.file to prevent the name and version
```

File: lib/lint.ml [hunk 6]
```diff
@@ -0,0 +1,310 @@
...
+     fields being automatically added *)
+  let ic = open_in opam_path in
```
Comment (shonfeder) [inline lib/lint.ml]: Probably should use [`In_channel.with_open_text`](https://ocaml.org/manual/5.2/api/In_channel.html#channels) (or `_bin`) to ensure the channel is closed, or else close it manually (along with the usual safety checks to ensure it is closed in case exceptions are raised).

File: lib/lint.ml [hunk 7]
```diff
@@ -0,0 +1,310 @@
...
+  let opam = OpamFile.OPAM.read_from_channel ic in
+  let errors = Checks.run_checks ~pkg opam in
+  match errors with
+  | [] -> print_endline "No errors"
+  | _ ->
+      errors |> List.map msg_of_error |> List.iter print_endline;
```
Comment (shonfeder) [inline lib/lint.ml]: Not that it will matter with the sizes we're dealing with, but still would be nice to avoid having to traverse the list twice with something like:

```suggestion
      errors |> List.iter (fun e -> e |> msg_of_error |> print_endline);
```

File: lib/lint.ml [hunk 8]
```diff
@@ -0,0 +1,310 @@
...
+         a bug in dune (https://github.com/ocaml/dune/issues/2147) this should \
+         never be the case. Please remove the {build} tag from its filter."
+        pkg
+
+let ( // ) = Filename.concat
+
+let path_from_pkg ~repo_dir pkg =
+  repo_dir // "packages"
+  // OpamPackage.Name.to_string (OpamPackage.name pkg)
+  // OpamPackage.to_string pkg
```
Comment (shonfeder) [inline lib/lint.ml]: Since `path_from_pkg` is only used once, IMO it would be cleaner to make this something like

```suggestion

let opam_file_path_from_pkg ~repo_dir pkg =
  let ( // ) = Filename.concat in
  repo_dir // "packages"
  // OpamPackage.Name.to_string (OpamPackage.name pkg)
  // OpamPackage.to_string pkg
  // "opam"
```

Which will then simplify line 300 a tad.

File: lib/lint.ml [hunk 9]
```diff
@@ -0,0 +1,310 @@
...
+    | UnmatchedVersion of OpamPackage.Version.t
+    | DubiousDuneSubst
+    | DuneProjectMissing
+    | DuneConstraintMissing
+    | DuneIsBuild
+    | BadDuneConstraint of string * string
+    | FailedToDownload of string
+    | WeakChecksum of string
+    | PinDepends
+    | ExtraFiles
```
Comment (shonfeder) [inline lib/lint.ml:16]: Seems like this is missing the prefix checks from https://github.com/ocurrent/opam-repo-ci/blob/b7fb27b10db7053b85aa2d3531fc86fa1516782e/lib/lint.ml#L44-L45  That may be intentional at this stage, but just making sure.

File: lib/lint.ml [hunk 10]
```diff
@@ -0,0 +1,310 @@
...
+  let check_no_extra_files ~pkg opam =
+    match OpamFile.OPAM.extra_files opam with
+    | None | Some [] -> []
+    | Some _ -> [ (pkg, ExtraFiles) ]
+
+  let with_temp_dir prefix f =
+    let dir = D.create_temp_dir prefix in
+    let res = f dir in
+    D.remove_dir dir;
+    res
```
Comment (shonfeder) [inline lib/lint.ml]: Seems like this should maybe be moved into `Dir_helpers`?

File: lib/lint.ml [hunk 11]
```diff
@@ -0,0 +1,310 @@
...
+    let is_build = ref false in
+    let rec get_lower_bound = function
+      | OpamFormula.Atom
+          (OpamTypes.Constraint ((`Gt | `Geq | `Eq), OpamTypes.FString version))
+        ->
+          Some version
+      | Atom (Filter (FIdent (_, var, _)))
+        when String.equal (OpamVariable.to_string var) "build" ->
+          is_build := true;
+          None (* TODO: remove this hack *)
```
Comment (shonfeder) [inline lib/lint.ml]: What's being hacked here?

File: lib/lint.ml [hunk 12]
```diff
@@ -0,0 +1,310 @@
...
+                (List [ Atom "lang"; Atom "dune"; Atom version ], _) ->
+                Ok (Some version)
+            | Done _ -> failwith "(lang dune ...) is not the first construct"
+            | Cont _ -> failwith "Failed to parse the dune-project file"
+          with e ->
+            close_in_noerr ch;
+            raise e
+        with
+        | Sys_error _ -> Ok None
+        | e -> raise e)
```
Comment (shonfeder) [inline lib/lint.ml]: I think this could probably be cleaned up using `In_channel.with_open_*` too? E.g. with something like

```suggestion
      let dune_project = Filename.concat dir "dune-project" in
      match In_channel.input_all |> In_channel.with_open_text dune_project |> Sexplib.Sexp.parse with
      | exception Sys_error _ -> Ok None
      | Sexplib.Sexp.Done (List [ Atom "lang"; Atom "dune"; Atom version ], _) ->
        Ok (Some version)
      | Done _ -> Error "(lang dune ...) is not the first construct"
      | Cont _ -> Error "Failed to parse the dune-project file"
```

**Review by shonfeder — COMMENTED** [review]
Excited to see this coming together! =D

File: lib/lint.ml (hunk 9 above)
Comment (punchagan) [reply lib/lint.ml:16]: Yes, that's intentional at this stage. There's a couple more (?) checks missing too, which I plan to add soon.
